### PR TITLE
K8SPXC-871: update finalizers in etcd for manually deleted pvc backup

### PR DESCRIPTION
**Pull request by defbin**

### CODE REVIEW

File: pkg/controller/pxcbackup/controller.go [hunk 1]
```diff
@@ -272,7 +272,18 @@ func (r *ReconcilePerconaXtraDBClusterBackup) tryRunS3BackupFinalizerJob(cr *api
...
 
 	if cr.Status.S3 == nil || cr.Status.Destination == "" || !strings.HasPrefix(cr.Status.Destination, "s3://") {
+		finalizersBefore := len(cr.GetFinalizers())
 		removeS3Finalizer(cr)
+
+		if finalizersBefore == len(cr.GetFinalizers()) {
+			return nil
+		}
+
+		if err := r.client.Update(context.TODO(), cr); err != nil {
```
Comment (delgod) [inline pkg/controller/pxcbackup/controller.go:282]: I am not sure maybe `r.client.Update` can be part of `removeS3Finalizer` function

**Review by delgod — APPROVED** [review]

File: pkg/controller/pxcbackup/controller.go (hunk 1 above)
Comment (defbin) [reply pkg/controller/pxcbackup/controller.go:282]: in the function below there are 
https://github.com/percona/percona-xtradb-cluster-operator/blob/daead24b10147355eaeca27e1c7e6b39e61cffa2/pkg/controller/pxcbackup/controller.go#L291
and
https://github.com/percona/percona-xtradb-cluster-operator/blob/daead24b10147355eaeca27e1c7e6b39e61cffa2/pkg/controller/pxcbackup/controller.go#L297

also it creates goroutine
https://github.com/percona/percona-xtradb-cluster-operator/blob/daead24b10147355eaeca27e1c7e6b39e61cffa2/pkg/controller/pxcbackup/controller.go#L291
that call the function
https://github.com/percona/percona-xtradb-cluster-operator/blob/daead24b10147355eaeca27e1c7e6b39e61cffa2/pkg/controller/pxcbackup/controller.go#L372

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | failed |
| upgrade-haproxy | passed |
| recreate | passed |
| big-data | passed |
| scaling | passed |
| scheduled-backup | passed |
| storage | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| init-deploy | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| self-healing-chaos | passed |
| limits | passed |
| demand-backup | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| smart-update | passed |
| affinity | passed |
| one-pod | passed |
| self-healing-advanced-chaos | passed |
| auto-tuning | passed |
| pitr | passed |
| proxysql-sidecar-res-limits | passed |
| operator-self-healing | passed |
| users | passed |
| operator-self-healing-chaos | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/987/commits/daead24b10147355eaeca27e1c7e6b39e61cffa2
image: `perconalab/percona-xtradb-cluster-operator:PR-987-daead24b`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| big-data | passed |
| scheduled-backup | passed |
| storage | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| self-healing-advanced | passed |
| upgrade-haproxy | passed |
| recreate | passed |
| self-healing-advanced-chaos | passed |
| restore-to-encrypted-cluster | passed |
| operator-self-healing | passed |
| demand-backup | passed |
| operator-self-healing-chaos | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| scaling | passed |
| scaling-proxysql | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/987/commits/daead24b10147355eaeca27e1c7e6b39e61cffa2
image: `perconalab/percona-xtradb-cluster-operator:PR-987-daead24b`

File: pkg/controller/pxcbackup/controller.go (hunk 1 above)
Comment (egegunes) [reply pkg/controller/pxcbackup/controller.go:282]: yes, I'd prefer the update in `removeS3Finalizer`

**Review by nonemax — DISMISSED** [review]

File: pkg/controller/pxcbackup/controller.go (hunk 1 above)
Comment (defbin) [reply pkg/controller/pxcbackup/controller.go]: moved to `removeS3Finalizer`. optimised memory usage

File: pkg/controller/pxcbackup/controller.go [hunk 2]
```diff
@@ -252,18 +252,27 @@ func (r *ReconcilePerconaXtraDBClusterBackup) Reconcile(request reconcile.Reques
...
-
-	for _, f := range cr.GetFinalizers() {
-		if f == api.FinalizerDeleteS3Backup {
-			continue
+func removeS3Finalizer(cl client.Client, cr *api.PerconaXtraDBClusterBackup) error {
+	finalizers := cr.GetFinalizers()
+	beforeLen := len(finalizers)
+
+	for i := 0; i != len(finalizers); i++ {
+		if finalizers[i] == api.FinalizerDeleteS3Backup {
```
Comment (egegunes) [inline pkg/controller/pxcbackup/controller.go:260]: I think the gains from the optimization don't compensate the loss of readability

File: pkg/controller/pxcbackup/controller.go [hunk 3]
```diff
@@ -252,18 +252,27 @@ func (r *ReconcilePerconaXtraDBClusterBackup) Reconcile(request reconcile.Reques
...
 
-		filteredFins = append(filteredFins, f)
+	if beforeLen == len(finalizers) {
+		return nil
 	}
 
-	cr.SetFinalizers(filteredFins)
+	cr.SetFinalizers(finalizers)
+	if err := cl.Update(context.TODO(), cr); err != nil {
+		return errors.Wrap(err, "remove S3 finalizer")
```
Comment (egegunes) [inline pkg/controller/pxcbackup/controller.go:272]: IMO it's better to describe the action that failed, rather than copying function's name

**Review by egegunes — CHANGES_REQUESTED** [review]

File: pkg/controller/pxcbackup/controller.go (hunk 2 above)
Comment (defbin) [reply pkg/controller/pxcbackup/controller.go:260]: but this is idiomatic way to remove item from slice

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/987/commits/8b271a7d9538f34c130128662b3b0d41a3209683
image: `perconalab/percona-xtradb-cluster-operator:PR-987-8b271a7d`

File: pkg/controller/pxcbackup/controller.go (hunk 3 above)
Comment (defbin) [reply pkg/controller/pxcbackup/controller.go:273]: I just did it to have error like `failed to run finalizers: try to run S3 backup finalizer job: remove S3 finalizer: [...root error...]`

but can change it to what you will suggest

Comment (egegunes) [reply pkg/controller/pxcbackup/controller.go:273]: in this case `update CR` would be better

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| storage | passed |
| scheduled-backup | passed |
| scaling | passed |
| recreate | passed |
| upgrade-haproxy | passed |
| big-data | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| security-context | passed |
| init-deploy | passed |
| demand-backup | passed |
| limits | passed |
| smart-update | passed |
| monitoring-2-0 | passed |
| self-healing-chaos | passed |
| affinity | passed |
| one-pod | passed |
| self-healing-advanced | passed |
| auto-tuning | passed |
| pitr | passed |
| self-healing-advanced-chaos | failed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| upgrade-consistency | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | failed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| demand-backup-encrypted-with-tls | failed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/987/commits/acec99e0fbe9a24fc9264958d95bc654212b798f
image: `perconalab/percona-xtradb-cluster-operator:PR-987-acec99e0`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| storage | passed |
| self-healing | passed |
| upgrade-proxysql | passed |
| self-healing-chaos | passed |
| smart-update | passed |
| self-healing-advanced | passed |
| self-healing-advanced-chaos | passed |
| operator-self-healing | passed |
| upgrade-consistency | passed |
| operator-self-healing-chaos | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/987/commits/39d53b830270f9b7479eacc0103722e37429c9e7
image: `perconalab/percona-xtradb-cluster-operator:PR-987-39d53b83`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| upgrade-haproxy | passed |
| storage | passed |
| scaling | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| self-healing-advanced | passed |
| self-healing-advanced-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| recreate | passed |
| scheduled-backup | passed |
| haproxy | failed |
| upgrade-proxysql | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| security-context | passed |
| init-deploy | passed |
| demand-backup | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| pitr | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| demand-backup-encrypted-with-tls | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/987/commits/39d53b830270f9b7479eacc0103722e37429c9e7
image: `perconalab/percona-xtradb-cluster-operator:PR-987-39d53b83`

**Review by egegunes — DISMISSED** [review]

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scheduled-backup | passed |
| upgrade-haproxy | passed |
| recreate | passed |
| haproxy | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| storage | passed |
| scaling | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| self-healing-chaos | passed |
| security-context | passed |
| self-healing-advanced | passed |
| self-healing-advanced-chaos | passed |
| big-data | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/987/commits/39d53b830270f9b7479eacc0103722e37429c9e7
image: `perconalab/percona-xtradb-cluster-operator:PR-987-39d53b83`

**Review by delgod — APPROVED** [review]

File: pkg/controller/pxcbackup/utils_test.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	"testing/quick"
+)
+
+type finalizer string
+
+func (finalizer) Generate(r *rand.Rand, size int) reflect.Value {
+	return reflect.ValueOf(finalizer(strconv.FormatInt(r.Int63(), 2)))
+}
+
+func TestRemoveStringsFromSlice(t *testing.T) {
```
Comment (egegunes) [inline pkg/controller/pxcbackup/utils_test.go:18]: I don't understand what are we testing here?

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| storage | passed |
| upgrade-haproxy | passed |
| recreate | passed |
| haproxy | failed |
| big-data | passed |
| scheduled-backup | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| init-deploy | passed |
| self-healing-chaos | passed |
| demand-backup | passed |
| smart-update | passed |
| self-healing-advanced | passed |
| limits | passed |
| self-healing-advanced-chaos | passed |
| monitoring-2-0 | passed |
| pitr | passed |
| affinity | passed |
| operator-self-healing | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| auto-tuning | passed |
| operator-self-healing-chaos | passed |
| demand-backup-encrypted-with-tls | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/987/commits/693189686ad2a5a06e3d2f26f1ae81b0983bef4b
image: `perconalab/percona-xtradb-cluster-operator:PR-987-69318968`

File: pkg/controller/pxcbackup/utils_test.go (hunk 4 above)
Comment (defbin) [reply pkg/controller/pxcbackup/utils_test.go:18]: the function
https://github.com/percona/percona-xtradb-cluster-operator/blob/693189686ad2a5a06e3d2f26f1ae81b0983bef4b/pkg/controller/pxcbackup/utils.go#L4
and that a caller can rely on the fn behaviour like here
https://github.com/percona/percona-xtradb-cluster-operator/blob/693189686ad2a5a06e3d2f26f1ae81b0983bef4b/pkg/controller/pxcbackup/controller.go#L260

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| scaling-proxysql | passed |
| security-context | passed |
| upgrade-haproxy | passed |
| storage | passed |
| scheduled-backup | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| self-healing-advanced | passed |
| self-healing-advanced-chaos | passed |
| big-data | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| haproxy | failed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/987/commits/693189686ad2a5a06e3d2f26f1ae81b0983bef4b
image: `perconalab/percona-xtradb-cluster-operator:PR-987-69318968`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| recreate | passed |
| storage | passed |
| restore-to-encrypted-cluster | passed |
| self-healing | passed |
| scheduled-backup | passed |
| demand-backup | passed |
| self-healing-chaos | passed |
| pitr | passed |
| self-healing-advanced | passed |
| haproxy | passed |
| demand-backup-encrypted-with-tls | passed |
| scaling | passed |
| self-healing-advanced-chaos | passed |
| scaling-proxysql | passed |
| operator-self-healing | passed |
| security-context | passed |
| operator-self-healing-chaos | passed |
| big-data | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/987/commits/693189686ad2a5a06e3d2f26f1ae81b0983bef4b
image: `perconalab/percona-xtradb-cluster-operator:PR-987-69318968`

**Review by cap1984 — DISMISSED** [review]

**Review by egegunes — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| scheduled-backup | passed |
| upgrade-haproxy | passed |
| storage | passed |
| recreate | passed |
| scaling | passed |
| big-data | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| security-context | passed |
| self-healing-chaos | passed |
| limits | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| affinity | passed |
| smart-update | passed |
| one-pod | passed |
| auto-tuning | passed |
| self-healing-advanced-chaos | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | passed |
| users | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| demand-backup-encrypted-with-tls | failed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/987/commits/bc1161ce8feab02641de73d7a31d3cc29b9bbee1
image: `perconalab/percona-xtradb-cluster-operator:PR-987-bc1161ce`

**Review by nonemax — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| upgrade-haproxy | passed |
| one-pod | passed |
| storage | passed |
| scaling | passed |
| auto-tuning | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| proxysql-sidecar-res-limits | passed |
| self-healing-chaos | passed |
| security-context | passed |
| users | passed |
| self-healing-advanced | passed |
| tls-issue-self | passed |
| big-data | passed |
| scheduled-backup | passed |
| self-healing-advanced-chaos | passed |
| tls-issue-cert-manager | passed |
| operator-self-healing | passed |
| tls-issue-cert-manager-ref | passed |
| operator-self-healing-chaos | passed |
| validation-hook | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/987/commits/bc1161ce8feab02641de73d7a31d3cc29b9bbee1
image: `perconalab/percona-xtradb-cluster-operator:PR-987-bc1161ce`

**Review by delgod — APPROVED** [review]

**Pull request merged**